src/connectivity/subgraphs.rs

```diff
@@ -0,0 +1,82 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use crate::StablePyGraph;
+use hashbrown::HashSet;
+use petgraph::stable_graph::NodeIndex;
+use petgraph::EdgeType;
+use std::cmp::max;
+use std::ops::Not;
+
+// Implemented after ``Simple`` from
+// "Enumerating Connected Induced Subgraphs: Improved Delay and Experimental Comparison".
+// Christian Komusiewicz and Frank Sommer. Some more runtime and spaceimprovements can be gained by
+// implementing the data structure defined for Lemma 4 (essentially a more efficient way of
+// tracking set ``x`` and ``p``)
+pub fn k_connected_subgraphs<Ty: EdgeType + Sync>(
+    graph: &StablePyGraph<Ty>,
+    k: usize,
+) -> Vec<Vec<usize>> {
+    let mut connected_subgraphs = Vec::new();
+    let mut graph = graph.clone();
+
+    while graph.node_count() >= max(k, 1) {
+        let mut p: HashSet<NodeIndex> = HashSet::new();
+        let v: NodeIndex = graph.node_indices().next().unwrap();
```

Comment:
I'd use [`while let`](https://doc.rust-lang.org/rust-by-example/flow_control/while_let.html) to avoid having the `.unwrap()`. Then you can come the `if` to break the loop early at the end